### PR TITLE
Removal of depreciated tauIDs

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

This PR is to remove depreciated BDT-based tauIDs superseded by deepTauIDs. The removal is preformed at AOD and miniAOD steps, nanoAOD is adjusted. Subset of old IDs is, however, kept for boosted taus as there is not viable replacement for them, yet. 
Expected changes:
* Removal of "byIsolationMVArun2" and "againstElectronMVA6" tau discriminators from AOD and RECO event content, and from slimmedTaus in miniAOD.

#### PR validation:

Matrix tests (`runTheMatrix.py -l limited -i all --ibeos) successful

### CODE REVIEW

**Comment by mariadalfonso** [comment]
please test

**Comment by mbluj** [comment]
The unit test fails due to problem with accessing input file - I checked and indeed the file is available only at tape. I can fix it, but need guidance which replacement should be used (preferably some file stored "permanently" for matrix tests or so).
What concerns differences in DQM they refer to removed tauIDs.

**Comment by jfernan2** [comment]
Thanks @mbluj 
The input file looks a bit old (94X) but I let @cms-sw/xpog-l2 to comment
About the DQM changes, with the removal of TauIDs, there are now (summary) plots which now get null entries, like:
https://tinyurl.com/ybjwzurd
and others where the number of labels in x-axis are reduced (since those IDs are not monitorized now):
https://tinyurl.com/ybtk3zx5

I wonder if this could be easily improved to avoid misunderstandings in the validation.
Thanks

**Comment by mbluj** [comment]
> Thanks @mbluj The input file looks a bit old (94X) but I let @cms-sw/xpog-l2 to comment About the DQM changes, with the removal of TauIDs, there are now (summary) plots which now get null entries, like: https://tinyurl.com/ybjwzurd and others where the number of labels in x-axis are reduced (since those IDs are not monitorized now): https://tinyurl.com/ybtk3zx5
> 
> I wonder if this could be easily improved to avoid misunderstandings in the validation. Thanks

Fixes to DQM other than purely technical, i.e. preventing it from failing, are beyond scope of this PR. The idea is to provide functional fixes, i.e. use of deepTauIDs in case of standard taus and MVAIso new/old DMs in case of boosted one, in other PR. This DQM PR is expected soon.

**Comment by jfernan2** [comment]
+1

**Comment by jpata** [comment]
the test `runtestPhysicsToolsNanoAOD` worked as recently as https://cmssdt.cern.ch/SDT/cgi-bin/logreader/cs8_amd64_gcc11/CMSSW_12_3_X_2022-02-28-2300/unitTestLogs/PhysicsTools/NanoAOD#
I wonder why it fails here now, perhaps @smuzaffar can help?

**Comment by smuzaffar** [comment]
the file is in ibeos cache `/eos/cms/store/user/cmsbuild/store/relval/CMSSW_9_4_0_pre3/RelValTTbar_13/MINIAODSIM/PU25ns_94X_mc2017_realistic_v4-v1/10000/52B94CC0-6FBB-E711-B577-0CC47A7C35F8.root` so it should have been read from ibeos cache. Mat be a glitch in eos/xrootd service is the reason behind this failure

File: PhysicsTools/NanoAOD/python/boostedTaus_cff.py [hunk 1]
```diff
@@ -9,9 +9,12 @@
 
 
 finalBoostedTaus = cms.EDFilter("PATTauRefSelector",
-    src = cms.InputTag("slimmedTausBoostedNewID"),
+    src = cms.InputTag("slimmedTausBoosted"),
     cut = cms.string("pt > 40 && tauID('decayModeFindingNewDMs') && (tauID('byVVLooseIsolationMVArun2017v2DBoldDMwLT2017') || tauID('byVVLooseIsolationMVArun2017v2DBoldDMdR0p3wLT2017') || tauID('byVVLooseIsolationMVArun2017v2DBnewDMwLT2017'))")
 )
+run3_nanoAOD_devel.toModify(finalBoostedTaus,
+                            cut = "pt > 40 && tauID('decayModeFindingNewDMs') && (tauID('byVVLooseIsolationMVArun2DBoldDMwLT') || tauID('byVVLooseIsolationMVArun2DBnewDMwLT'))"
+)
```
Comment (jpata) [inline PhysicsTools/NanoAOD/python/boostedTaus_cff.py:17]: I'm not sure if I understand why the Run3 modifier is needed. Shouldn't the selection on the new IDs be the new default for all NANO?

File: PhysicsTools/NanoAOD/python/boostedTaus_cff.py [hunk 2]
```diff
@@ -36,21 +39,37 @@
...
 )
 
-boostedTauTable.variables = _boostedTauVarsBase
-
+boostedTauTable.variables = cms.PSet(
+    _boostedTauVarsBase,
+    _boostedTauVarsMVAIso,
+    _boostedTauVarsAntiEleMVA
+)
+run3_nanoAOD_devel.toModify(boostedTauTable.variables,
```
Comment (jpata) [inline PhysicsTools/NanoAOD/python/boostedTaus_cff.py:65]: same comment here. shouldn't it be the new default, rather than a special config for Run3?

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 3]
```diff
@@ -348,8 +357,16 @@ def nanoAOD_customizeCommon(process):
...
-    (~(run2_nanoAOD_94XMiniAODv1 | run2_nanoAOD_94X2016 | run2_nanoAOD_94XMiniAODv2 | run2_nanoAOD_102Xv1 | run2_nanoAOD_106Xv1)).toModify(process, lambda p : nanoAOD_addBoostedTauIds(p))
+    nanoAOD_tau_switch = cms.PSet(
+        idsToAdd = cms.vstring()
+    )
+    (run2_nanoAOD_94XMiniAODv1 | run2_nanoAOD_94X2016 | run2_nanoAOD_94XMiniAODv2 | run2_nanoAOD_102Xv1 | run2_nanoAOD_106Xv1).toModify(nanoAOD_tau_switch, idsToAdd = ["deepTau2017v2p1"])
+    (run2_nanoAOD_94XMiniAODv1 | run2_nanoAOD_94X2016 | run2_nanoAOD_94XMiniAODv2 | run2_nanoAOD_102Xv1 | run2_nanoAOD_106Xv1).toModify(process, lambda p : nanoAOD_addTauIds(p, nanoAOD_tau_switch.idsToAdd.value()))
+    nanoAOD_boostedTau_switch = cms.PSet(
+        idsToAdd = cms.vstring("2017v2", "dR0p32017v2", "newDM2017v2","againstEle2018")
+    )
+    run3_nanoAOD_devel.toModify(nanoAOD_boostedTau_switch, idsToAdd = [])
```
Comment (jpata) [inline PhysicsTools/NanoAOD/python/nano_cff.py:368]: same comment here

**Review by jpata — COMMENTED** [review]
As I understand, if a new default is introduced, it should be without modifiers for a specific period.

File: PhysicsTools/NanoAOD/python/boostedTaus_cff.py (hunk 1 above)
Comment (jpata) [reply PhysicsTools/NanoAOD/python/boostedTaus_cff.py:17]: supposing someone was to produce NANO for Run2 again, should the new or old setting be used in this case?

**Comment by mariadalfonso** [comment]
@run3_nanoAOD_devel should not be used indeed.

first the ID run3 should be in the mini and then nano will just read. This is not the case and should be fixed.
The ID need to re-run only when we read the miniv2 of UL.

**Comment by easilar** [comment]
> Thanks @mbluj The input file looks a bit old (94X) but I let @cms-sw/xpog-l2 to comment About the DQM changes, with the removal of TauIDs, there are now (summary) plots which now get null entries, like: https://tinyurl.com/ybjwzurd and others where the number of labels in x-axis are reduced (since those IDs are not monitorized now): https://tinyurl.com/ybtk3zx5
> 
> I wonder if this could be easily improved to avoid misunderstandings in the validation. Thanks

There will be a new PR soon also for Tau DQM to update the IDs and also move from AOD to miniAOD. The PR is in preparation. In fact at the moment I am running the runthematrix for PR. Once it is successfully done we will make a PR that hopefully will cure possible empty histograms.

**Comment by mbluj** [comment]
> @run3_nanoAOD_devel should not be used indeed.
> 
> first the ID run3 should be in the mini and then nano will just read. This is not the case and should be fixed. The ID need to re-run only when we read the miniv2 of UL, specifying the modifier run2_nanoAOD_106Xv2. For the previous eras, this boosted-tau tasks was already off.

Thank you for your comments. I will proceed along these lines.

**Comment by mariadalfonso** [comment]
please test

**Comment by mbluj** [comment]
It seems that some NanoAOD DQM workflows use default configuration even with old input files for eras which are not run in the production workflow, but it could be than I am wrong. I plan to take a closer look tomorrow.

**Comment by gouskos** [comment]
please test

**Comment by mbluj** [comment]
@gouskos I appologise, but after visual inspection with web interface I found a small issue, so I pushed again. Could you please restart the tests?

**Comment by gouskos** [comment]
> @gouskos I appologise, but after visual inspection with web interface I found a small issue, so I pushed again. Could you please restart the tests?

No problem @mbluj !

**Comment by gouskos** [comment]
please abort

**Comment by mbluj** [comment]
@gouskos thank you. Now it should be fine, so after automatic checks the tests can be started again.

**Comment by clacaputo** [comment]
@cmsbuild please test

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 4]
```diff
@@ -348,8 +357,16 @@ def nanoAOD_customizeCommon(process):
...
+    nanoAOD_tau_switch = cms.PSet(
+        idsToAdd = cms.vstring()
+    )
+    (run2_nanoAOD_94XMiniAODv1 | run2_nanoAOD_94X2016 | run2_nanoAOD_94XMiniAODv2 | run2_nanoAOD_102Xv1 | run2_nanoAOD_106Xv1).toModify(nanoAOD_tau_switch, idsToAdd = ["deepTau2017v2p1"])
+    (run2_nanoAOD_94XMiniAODv1 | run2_nanoAOD_94X2016 | run2_nanoAOD_94XMiniAODv2 | run2_nanoAOD_102Xv1 | run2_nanoAOD_106Xv1).toModify(process, lambda p : nanoAOD_addTauIds(p, nanoAOD_tau_switch.idsToAdd.value()))
+    nanoAOD_boostedTau_switch = cms.PSet(
+        idsToAdd = cms.vstring()
+    )
+    run2_nanoAOD_106Xv2.toModify(nanoAOD_boostedTau_switch, idsToAdd = ["2017v2", "dR0p32017v2", "newDM2017v2","againstEle2018"])
+    (~(run2_nanoAOD_94XMiniAODv1 | run2_nanoAOD_94X2016 | run2_nanoAOD_94XMiniAODv2 | run2_nanoAOD_102Xv1 | run2_nanoAOD_106Xv1)).toModify(process, lambda p : nanoAOD_addBoostedTauIds(p, nanoAOD_boostedTau_switch.idsToAdd.value()))
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/nano_cff.py:369]: this (~manyeras) can be replaced with the 

run2_nanoAOD_106Xv2.toModify(process, lambda p : nanoAOD_addBoostedTauIds(p, nanoAOD_boostedTau_switch.idsToAdd.value()))

This is the only mini that doesn't contain the default ID and one need to re-run.

Comment (mbluj) [reply PhysicsTools/NanoAOD/python/nano_cff.py:369]: Yes, right. I will change it as proposed.

**Comment by mariadalfonso** [comment]
> * Reco comparison results: 3542 differences found in the comparisons

none of these are related to the nano.
i.e. no difference in wf 1325.81 (106Xv1), 136.8523 (106Xv2)

**Comment by mariadalfonso** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by clacaputo** [comment]
Reco changes in AOD are in line with the PR, while differences in miniAOD can be related to a known issue in the validate script https://github.com/cms-sw/cmssw/pull/36179#issuecomment-1015371027.
I'm checking them locally.

**Comment by clacaputo** [comment]
I've checked for wf `10024.0` the differences in miniAOD for `tauIDs`. Let's take the `slimmedTausBoosted__RECO_obj___tauIDs__19` shown in the picture below

all_mini_OldVSNew_TTbar13TeV2017wf10024p0c_patTaus_slimmedTausBoosted__RECO_obj___tauIDs__19__second

Comparing `pT, eta, tauIDs[19], tauIDs[3]` for `PR+CMSSW_12_3_X_2022-03-04-2300` and `CMSSW_12_3_X_2022-03-04-2300`, we found that `tauIDs[19]` point to a different ID ( `byIsolationMVArun2DBnewDMwLTraw` Vs `byIsolationMVArun2v1DBdR03oldDMwLTraw`, see results below), while `pT, eta, tauIDs[3]` are the same

**PR+CMSSW_12_3_X_2022-03-04-2300** 
```bash
tau pT   tau eta         tauIDs[19]                                                                            tauIDs[3]
77.55    -1.73   { "byIsolationMVArun2DBnewDMwLTraw", -0.990194f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 24.7089f }
38.81    -1.18   { "byIsolationMVArun2DBnewDMwLTraw", -0.947184f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 39.1803f }
29.63    -0.18   { "byIsolationMVArun2DBnewDMwLTraw", -0.884993f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 26.3342f }
19.34    -0.65   { "byIsolationMVArun2DBnewDMwLTraw", -0.980453f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 60.0140f }
36.12    -0.96   { "byIsolationMVArun2DBnewDMwLTraw", -0.991541f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 33.4097f }
36.42    -1.40   { "byIsolationMVArun2DBnewDMwLTraw", -0.568650f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 1.94070f }
105.37   0.57    { "byIsolationMVArun2DBnewDMwLTraw", -0.768153f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 13.4821f }
63.61    -0.18   { "byIsolationMVArun2DBnewDMwLTraw", -0.940415f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 22.8685f }
50.33    1.27    { "byIsolationMVArun2DBnewDMwLTraw", -0.0154546f }      { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 7.33329f }
32.36    1.25    { "byIsolationMVArun2DBnewDMwLTraw", -0.962710f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 10.9408f }
18.27    -1.33   { "byIsolationMVArun2DBnewDMwLTraw", -0.985566f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 46.5323f }
53.44    -0.92   { "byIsolationMVArun2DBnewDMwLTraw", -0.956164f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 27.0370f }
28.85    -0.90   { "byIsolationMVArun2DBnewDMwLTraw", -0.911898f }       { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 22.0325f }
```

**CMSSW_12_3_X_2022-03-04-2300**
```bash
tau pT   tau eta         tauIDs[19]                                                                              tauIDs[3]
77.55    -1.73   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -1.00000f }  { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 24.7089f }
38.81    -1.18   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -1.00000f }  { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 39.1803f }
29.63    -0.18   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -1.00000f }  { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 26.3342f }
19.34    -0.65   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -1.00000f }  { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 60.0140f }
36.12    -0.96   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -1.00000f }  { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 33.4097f }
36.42    -1.40   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -0.642849f }         { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 1.94070f }
105.37   0.57    { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -0.968865f }         { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 13.4821f }
63.61    -0.18   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -0.974198f }         { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 22.8685f }
50.33    1.27    { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -0.436733f }         { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 7.33329f }
32.36    1.25    { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -1.00000f }  { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 10.9408f }
18.27    -1.33   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -0.868038f }         { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 46.5323f }
53.44    -0.92   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -1.00000f }  { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 27.0370f }
28.85    -0.90   { "byIsolationMVArun2v1DBdR03oldDMwLTraw", -0.956641f }         { "byCombinedIsolationDeltaBetaCorrRaw3Hits", 22.0325f }
```

Summing up, the differences in miniAOD are in line with the PR

**Comment by clacaputo** [comment]
+reconstruction

- the PR remove depreciated BDT-based tauIDs superseded by deepTauIDs. A subset of old IDs is, however, kept for boosted taus
- reco changes are in line with the PR, see https://github.com/cms-sw/cmssw/pull/37091#issuecomment-1061005914

**Comment by mariadalfonso** [comment]
@mbluj 

there is some change in the idAntiEle2018 for the modifier run2_nanoAOD_106Xv2
is this intentional ?

Screen Shot 2022-03-07 at 23 21 57

**Comment by mbluj** [comment]
> @mbluj
> 
> there is some change in the idAntiEle2018 for the modifier run2_nanoAOD_106Xv2 is this intentional ?
> 
> Screen Shot 2022-03-07 at 23 21 57

In the previous version anti-e tauID was rerun,on top of miniAOD but not used. Instead anti-e from miniAOD was read. I fixed it, i.e. now the rerun ID is used which, I think, is a correct way. However, the difference should be small and I have not strong opinion what is better.

**Comment by mariadalfonso** [comment]
> > @mbluj
> > there is some change in the idAntiEle2018 for the modifier run2_nanoAOD_106Xv2 is this intentional ?
> > Screen Shot 2022-03-07 at 23 21 57
> 
> In the previous version anti-e tauID was rerunon top of miniAOD, but not used. Instead anti-e from miniAOD was read. I fixed it, i.e. now the rerun ID is used which, I think, is a correct implementation. However, the difference between the two IDs should be small and I have not strong opinion what is better read from mini or rerun.

ok, then let's consider it as bug fix for the future Run2 nano (that will read the same mini as specified by the run2_nanoAOD_106Xv2 )

**Comment by mariadalfonso** [comment]
+xpog

updates in view of Run3 + fix as in https://github.com/cms-sw/cmssw/pull/37091#issuecomment-1061590869

**Comment by perrotta** [comment]
+1
